src/Microsoft.Identity.Web.TokenAcquisition/ConfidentialClientApplicationBuilderExtension.cs

```diff
@@ -58,51 +96,54 @@ public static async Task<ConfidentialClientApplicationBuilder> WithClientCredent
                             else
                             {
                                 Logger.UsingManagedIdentity(logger);
-                                return builder.WithClientAssertion((credential.CachedValue as ManagedIdentityClientAssertion)!.GetSignedAssertionAsync);
+                                return credential;
                             }
                         }
                         if (credential.SourceType == CredentialSource.SignedAssertionFilePath)
                         {
                             if (!credential.Skip)
                             {
                                 Logger.UsingPodIdentityFile(logger, credential.SignedAssertionFileDiskPath ?? "not found");
-                                return builder.WithClientAssertion((credential.CachedValue as AzureIdentityForKubernetesClientAssertion)!.GetSignedAssertionAsync);
+                                return credential;
                             }
                         }
                         if (credential.SourceType == CredentialSource.SignedAssertionFromVault)
                         {
                             if (!credential.Skip)
                             {
                                 Logger.UsingSignedAssertionFromVault(logger, credential.KeyVaultUrl ?? "undefined");
-                                return builder.WithClientAssertion((credential.CachedValue as ClientAssertionProviderBase)!.GetSignedAssertionAsync);
+                                return credential;
                             }
                         }
                     }
 
                     if (credential.CredentialType == CredentialType.Certificate)
                     {
-                        if (credential.Certificate !=null)
+                        if (credential.Certificate != null)
                         {
                             Logger.UsingCertThumbprint(logger, credential.Certificate.Thumbprint);
-                            return builder.WithCertificate(credential.Certificate);
+                            return credential;
                         }
                     }
 
                     if (credential.CredentialType == CredentialType.Secret)
                     {
-                        return builder.WithClientSecret(credential.ClientSecret);
+                        return credential;
                     }
                 }
             }
 
-            if (clientCredentials.Any(c => c.CredentialType == CredentialType.Certificate))
+            if (clientCredentials.Any(c => c.CredentialType == CredentialType.Certificate || c.CredentialType == CredentialType.SignedAssertion))
```

Comment:
I changed this condition. I originally wanted to throw if no credential could be set, but this is not as simple because there are so MANY ways to set a secret (ASP options, CCA options etc.)